packages/core/contracts/optimistic-oracle-v2/implementation/WhitelistOptimisticOracleV2.sol

```diff
@@ -0,0 +1,413 @@
+     * @param requester sender of the initial price request.
+     * @param identifier price identifier to identify the existing request.
+     * @param ancillaryData ancillary data of the price being requested.
+     * @param whitelist address of the whitelist to set.
+     */
+    function requestManagerSetProposerWhitelist(
+        address requester,
+        bytes32 identifier,
+        bytes memory ancillaryData,
+        address whitelist
+    ) external nonReentrant() onlyRequestManager() {
+        bytes32 requestId = _getId(requester, identifier, 0, ancillaryData);
+        customProposerWhitelists[requestId] = AddressWhitelistInterface(whitelist);
+        emit CustomProposerWhitelistSet(requestId, requester, identifier, ancillaryData, whitelist);
+    }
+
+    /**
+     * @notice Proposes a price value on another address' behalf. Note: this address will receive any rewards that come
+     * from this proposal. However, any bonds are pulled from the caller.
+     * @dev Timestamp is omitted from the whitelist key derivation, so it would also apply for repeated requests.
+     * @param proposer address to set as the proposer.
+     * @param requester sender of the initial price request.
+     * @param identifier price identifier to identify the existing request.
+     * @param timestamp timestamp to identify the existing request.
+     * @param ancillaryData ancillary data of the price being requested.
+     * @param proposedPrice price being proposed.
+     * @return totalBond the amount that's pulled from the caller's wallet as a bond. The bond will be returned to
+     * the proposer once settled if the proposal is correct.
+     */
+    function proposePriceFor(
+        address proposer,
+        address requester,
+        bytes32 identifier,
+        uint256 timestamp,
+        bytes memory ancillaryData,
+        int256 proposedPrice
+    ) public override returns (uint256 totalBond) {
+        AddressWhitelistInterface whitelist = customProposerWhitelists[_getId(requester, identifier, 0, ancillaryData)];
+        if (address(whitelist) == address(0)) {
+            whitelist = defaultProposerWhitelist;
+        }
+
+        require(whitelist.isOnWhitelist(proposer) || msg.sender == owner(), "Proposer not whitelisted");
+        return super.proposePriceFor(proposer, requester, identifier, timestamp, ancillaryData, proposedPrice);
+    }
+
+    /**
+     * @notice Gets the custom proposer whitelist for a request.
+     * @dev This omits the timestamp from the key derivation, so the whitelist might have been set in advance of the
+     * request.
+     * @param requester sender of the initial price request.
+     * @param identifier price identifier to identify the existing request.
+     * @param ancillaryData ancillary data of the price being requested.
+     * @return AddressWhitelistInterface the custom proposer whitelist for the request or zero address if not set.
+     */
+    function getCustomProposerWhitelist(
```

Comment:
we should also extend this to have a function recursing the effective whitelist array and boolean on whether the whitelist is enabled.